prow/gerrit/adapter/adapter.go

```diff
@@ -91,38 +91,58 @@ func NewController(ctx context.Context, prowJobClient prowv1.ProwJobInterface, o
 	if err != nil {
 		logrus.WithError(err).Fatal("Error creating gerrit client.")
 	}
-	gerritClient.Authenticate(cookiefilePath, tokenPathOverride)
-
 	c := &Controller{
 		prowJobClient:      prowJobClient,
 		config:             cfg,
 		gc:                 gerritClient,
 		tracker:            lastSyncTracker,
 		projectsOptOutHelp: projectsOptOutHelpMap,
 	}
+
+	// applyGlobalConfig reads gerrit configurations from global gerrit config,
+	// it will completely override previously configured gerrit hosts and projects.
+	// it will also by the way authenticate gerrit
+	c.applyGlobalConfig(cfg, gerritClient, lastSyncTracker, cookiefilePath, tokenPathOverride)
+
+	// Authenticate creates a goroutine for rotating token secrets when called the first
+	// time, afterwards it only authenticate once.
+	// applyGlobalConfig calls authenticate only when global gerrit config presents,
+	// call it here is required for cases where gerrit repos are defined as command
+	// line arg(which is going to be deprecated).
+	gerritClient.Authenticate(cookiefilePath, tokenPathOverride)
+
+	return c
+}
+
+func (c *Controller) applyGlobalConfig(cfg config.Getter, gerritClient *client.Client, lastSyncTracker *syncTime, cookiefilePath, tokenPathOverride string) {
+	c.applyGlobalConfigOnce(cfg, gerritClient, lastSyncTracker, cookiefilePath, tokenPathOverride)
+
 	go func() {
 		for {
-			orgReposConfig := cfg().Gerrit.OrgReposConfig
-			if orgReposConfig == nil {
-				time.Sleep(time.Second)
-				continue
-			}
-			// Use globally defined gerrit repos if present
-			if err := gerritClient.UpdateClients(orgReposConfig.AllRepos()); err != nil {
-				logrus.WithError(err).Error("Updating clients.")
-			}
-			if err := lastSyncTracker.update(orgReposConfig.AllRepos()); err != nil {
-				logrus.WithError(err).Error("Syncing states.")
-			}
-			// Updates a map, lock to make sure it's thread safe.
-			c.lock.Lock()
-			c.projectsOptOutHelp = orgReposConfig.OptOutHelpRepos()
-			c.lock.Unlock()
+			c.applyGlobalConfigOnce(cfg, gerritClient, lastSyncTracker, cookiefilePath, tokenPathOverride)
 			// No need to spin constantly, give it a break. It's ok that config change has one second delay.
 			time.Sleep(time.Second)
 		}
 	}()
-	return c
+}
+
+func (c *Controller) applyGlobalConfigOnce(cfg config.Getter, gerritClient *client.Client, lastSyncTracker *syncTime, cookiefilePath, tokenPathOverride string) {
+	orgReposConfig := cfg().Gerrit.OrgReposConfig
+	if orgReposConfig == nil {
+		return
+	}
+	// Use globally defined gerrit repos if present
+	if err := gerritClient.UpdateClients(orgReposConfig.AllRepos()); err != nil {
+		logrus.WithError(err).Error("Updating clients.")
+	}
+	if err := lastSyncTracker.update(orgReposConfig.AllRepos()); err != nil {
+		logrus.WithError(err).Error("Syncing states.")
+	}
```

Comment:
I think we still need to include an `Authenticate` call some time after `UpdateClients` otherwise we only ever call `Authenticate` once.

Reply:
Authenticate spins off a goroutine updating when token refreshes, which is the only case where authenticate needs to be rerun when global config is not present

Reply:
When global config is present and we add a new client to the `handler` map we need to call this line before using the handler, right? https://github.com/kubernetes/test-infra/blob/fcb86429c4b1f332290b2e000e58f4c6e7108b6c/prow/gerrit/client/client.go#L202

Reply:
ah, couldn't believe that I had deleted the `Authenticate` call below from original change! Awesome catch!